src/synchronous_product/include/synchronous_product/search_tree.h

```diff
@@ -65,6 +68,115 @@ struct SearchTreeNode
+	 * the node can be labelled "GOOD". The call should be propagated to the parent node in case the
+	 * labelling has been determined.
+	 * @param controller_actions The set of controller actions
+	 * @param environment_actions The set of environment actions
+	 */
+	void
+	label_propagate(const std::set<ActionType> &controller_actions,
+	                const std::set<ActionType> &environment_actions)
+	{
+		// leaf-nodes should always be labelled directly
+		assert(!children.empty() || label != NodeLabel::UNLABELED);
+		// if not already happened: call recursively on parent node
+		if (children.empty()) {
+			assert(label != NodeLabel::UNLABELED);
+			if (parent != nullptr) {
+				SPDLOG_TRACE("Node {} is a leaf, propagate labels.", *this);
+				parent->label_propagate(controller_actions, environment_actions);
+			}
+		}
+		// do nothing if the node is already labelled
+		if (label != NodeLabel::UNLABELED && !children.empty()) {
+			SPDLOG_TRACE("Node is already labelled, abort.");
+			return;
+		}
+		// find good and bad child nodes which are already labelled and determine their order (with
+		// respect to time). Also keep track of yet unlabelled nodes.
+		RegionIndex first_good_controller_step{std::numeric_limits<RegionIndex>::max()};
+		RegionIndex first_non_good_environment_step{std::numeric_limits<RegionIndex>::max()};
+		RegionIndex first_bad_environment_step{std::numeric_limits<RegionIndex>::max()};
+		for (const auto &child : children) {
+			for (const auto &[step, action] : child->incoming_actions) {
+				if (child->label == NodeLabel::TOP
+				    && controller_actions.find(action) != std::end(controller_actions)) {
+					first_good_controller_step = std::min(first_good_controller_step, step);
+				} else if (child->label == NodeLabel::BOTTOM
+				           && environment_actions.find(action) != std::end(environment_actions)) {
+					first_bad_environment_step = std::min(first_bad_environment_step, step);
+				} else if (child->label == NodeLabel::UNLABELED
+				           && environment_actions.find(action) != std::end(environment_actions)) {
+					first_non_good_environment_step = std::min(first_non_good_environment_step, step);
+				}
+			}
+		}
+		SPDLOG_TRACE("First good controller step at {}, first non-good env. action step at {}, first "
+		             "bad env. action at {}",
+		             first_good_controller_step,
+		             first_non_good_environment_step,
+		             first_bad_environment_step);
+		// cases in which incremental labelling can be applied and recursive calls should be issued
+		if (first_good_controller_step < first_non_good_environment_step
+		    && first_good_controller_step < first_bad_environment_step) {
+			label = NodeLabel::TOP;
+			SPDLOG_TRACE("Label with TOP");
+			if (parent != nullptr) {
+				parent->label_propagate(controller_actions, environment_actions);
+			}
+		} else if (first_bad_environment_step < first_good_controller_step) {
```

Comment:
What if there is a good controller step but the respective child has not been labeled yet? I think we also need to track the `first_non_bad_controller_step` to cover that case

Reply:
Hm, good point. This becomes more relevant in the case of multithreaded tree expansion, currently in the purelty sequential setting nodes are unfolded in chronological order. Which in retrospect renders the "non_good_environmental" action useless, I think. I have added this.